File: app/navitia/journey_parser.py
"""
Navitia API response parser.

Extracts structured journey data from raw Navitia JSON responses.
No I/O or network calls — safe to use in any context.
"""

from dataclasses import dataclass, field


@dataclass
class JourneyResult:
    """
    A structured representation of one train journey from Navitia.

    Attributes:
        from_station_nom: Name of the departure station.
        to_station_nom: Name of the arrival station.
        departure_datetime: ISO 8601 departure datetime string (e.g. '2026-04-09T08:04:00').
        arrival_datetime: ISO 8601 arrival datetime string.
        duration_minutes: Total journey duration in minutes.
        nb_transfers: Number of transfers in the journey.
        sections: List of simplified section summaries (mode, from, to, duration_min).
    """

    from_station_nom: str
    to_station_nom: str
    departure_datetime: str
    arrival_datetime: str
    duration_minutes: int
    nb_transfers: int
    sections: list[dict] = field(default_factory=list)


def _navitia_datetime_to_iso(navitia_dt: str) -> str:
    """
    Convert a Navitia datetime string to ISO 8601 format.

    Navitia uses the format '20260409T080400' (no separators).
    ISO 8601 format is '2026-04-09T08:04:00'.

    Args:
        navitia_dt: Datetime string in Navitia format 'YYYYMMDDTHHmmss'.

    Returns:
        ISO 8601 string, or the original string if it cannot be parsed.
    """
    try:
        date_part = navitia_dt[:8]
        time_part = navitia_dt[9:15] if len(navitia_dt) >= 15 else navitia_dt[9:]
        year = date_part[:4]
        month = date_part[4:6]
        day = date_part[6:8]
        hour = time_part[:2]
        minute = time_part[2:4]
        second = time_part[4:6] if len(time_part) >= 6 else "00"
        return f"{year}-{month}-{day}T{hour}:{minute}:{second}"
    except (IndexError, ValueError):
        return navitia_dt


def _extract_station_name(place: dict) -> str:
    """
    Extract a human-readable station name from a Navitia 'place' dict.

    Args:
        place: A Navitia place object (from journey 'from' or 'to' fields,
               or section endpoint).

    Returns:
        Station name string. Returns empty string if the structure is unexpected.
    """
    stop_area = place.get("stop_area") or place.get("stop_point", {})
    if stop_area:
        return stop_area.get("name", "")
    return place.get("name", "")


def parse_journey_sections(journey: dict) -> list[dict]:
    """
    Extract simplified section summaries from a single Navitia journey dict.

    Each section is simplified to: mode, from station name, to station name,
    and duration in minutes.

    Args:
        journey: A single journey object from a Navitia API response.

    Returns:
        List of dicts with keys 'mode', 'from', 'to', 'duration_min'.
        Empty list if no sections are present.
    """
    sections = []
    for section in journey.get("sections", []):
        section_type = section.get("type", "")
        mode = section.get("display_informations", {}).get("physical_mode", section_type)
        from_place = _extract_station_name(section.get("from", {}))
        to_place = _extract_station_name(section.get("to", {}))
        duration_sec = section.get("duration", 0)
        sections.append({
            "mode": mode,
            "from": from_place,
            "to": to_place,
            "duration_min": round(duration_sec / 60),
        })
    return sections


def parse_best_journey(api_response: dict) -> JourneyResult | None:
    """
    Extract the best (first) journey from a Navitia journeys API response.

    Navitia returns journeys sorted by optimality (fastest / fewest transfers).
    This function takes the first journey in the list as the recommended option.

    Args:
        api_response: Parsed JSON dict from a Navitia /journeys endpoint call.

    Returns:
        JourneyResult for the first journey, or None if:
        - The response contains no journeys.
        - The 'journeys' key is missing.

    Raises:
        ValueError: If the response structure is unexpectedly malformed (missing
                    required fields on a journey that does exist).
    """
    journeys = api_response.get("journeys")
    if not journeys:
        return None

    journey = journeys[0]

    try:
        departure_dt = _navitia_datetime_to_iso(journey["departure_date_time"])
        arrival_dt = _navitia_datetime_to_iso(journey["arrival_date_time"])
        duration_sec = journey["duration"]
        nb_transfers = journey.get("nb_transfers", 0)
    except KeyError as exc:
        raise ValueError(f"Journey missing required field: {exc}") from exc

    from_nom = _extract_station_name(journey.get("sections", [{}])[0].get("from", {}))
    to_nom = _extract_station_name(
        journey.get("sections", [{}])[-1].get("to", {})
        if journey.get("sections")
        else {}
    )
    sections = parse_journey_sections(journey)

    return JourneyResult(
        from_station_nom=from_nom,
        to_station_nom=to_nom,
        departure_datetime=departure_dt,
        arrival_datetime=arrival_dt,
        duration_minutes=round(duration_sec / 60),
        nb_transfers=nb_transfers,
        sections=sections,
    )


def format_duration_minutes(total_minutes: int) -> str:
    """
    Convert a duration in minutes to a human-readable French-style string.

    Args:
        total_minutes: Non-negative integer number of minutes.

    Returns:
        Formatted string like '1h 35min' or '45min' (if under 1 hour)
        or '2h' (if exactly on the hour).

    Raises:
        ValueError: If total_minutes is negative.
    """
    if total_minutes < 0:
        raise ValueError(f"total_minutes must be non-negative, got {total_minutes}")
    hours = total_minutes // 60
    minutes = total_minutes % 60
    if hours == 0:
        return f"{minutes}min"
    if minutes == 0:
        return f"{hours}h"
    return f"{hours}h {minutes}min"

File: static/js/journey_parser.js
/**
 * journey_parser.js — Navitia API response parser (port of app/navitia/journey_parser.py).
 *
 * Extracts structured journey data from raw Navitia JSON responses.
 * No I/O or network calls — pure data transformation.
 *
 * Exposes: window.InterJourney
 */

(function () {
  "use strict";

  /**
   * Convert a Navitia datetime string to ISO 8601 format.
   *
   * Navitia format: '20260409T080400' (no separators).
   * ISO 8601 output: '2026-04-09T08:04:00'.
   *
   * @param {string} navitiaDatetime - Datetime string in Navitia format.
   * @returns {string} ISO 8601 string, or the original if conversion fails.
   */
  function navitiaDatetimeToIso(navitiaDatetime) {
    try {
      const year   = navitiaDatetime.slice(0, 4);
      const month  = navitiaDatetime.slice(4, 6);
      const day    = navitiaDatetime.slice(6, 8);
      const hour   = navitiaDatetime.slice(9, 11);
      const minute = navitiaDatetime.slice(11, 13);
      const second = navitiaDatetime.length >= 15 ? navitiaDatetime.slice(13, 15) : "00";
      return `${year}-${month}-${day}T${hour}:${minute}:${second}`;
    } catch (_) {
      return navitiaDatetime;
    }
  }

  /**
   * Extract a human-readable station name from a Navitia 'place' object.
   *
   * @param {Object} place - Navitia place object from a journey section endpoint.
   * @returns {string} Station name, or empty string if structure is unexpected.
   */
  function extractStationName(place) {
    if (!place) return "";
    const stopArea = place.stop_area || place.stop_point;
    if (stopArea) return stopArea.name || "";
    return place.name || "";
  }

  /**
   * Extract simplified section summaries from a single Navitia journey.
   *
   * @param {Object} journey - A single journey object from a Navitia API response.
   * @returns {Array<{mode: string, from: string, to: string, duration_min: number}>}
   *   Section summaries. Empty array if no sections are present.
   */
  function parseJourneySections(journey) {
    const sections = [];
    for (const section of (journey.sections || [])) {
      const sectionType = section.type || "";
      const mode = (section.display_informations && section.display_informations.physical_mode) || sectionType;
      const from = extractStationName(section.from);
      const to   = extractStationName(section.to);
      const durationMin = Math.round((section.duration || 0) / 60);
      sections.push({ mode, from, to, duration_min: durationMin });
    }
    return sections;
  }

  /**
   * Extract the best (first) journey from a Navitia journeys API response.
   *
   * Navitia returns journeys sorted by optimality. This function returns the
   * first entry as the recommended option.
   *
   * @param {Object} apiResponse - Parsed JSON from a Navitia /journeys call.
   * @returns {Object|null} Journey result object, or null if:
   *   - The response has no journeys.
   *   - The 'journeys' key is missing.
   *
   * The returned object has keys:
   *   from_station_nom, to_station_nom, departure_datetime, arrival_datetime,
   *   duration_minutes, nb_transfers, sections.
   */
  function parseBestJourney(apiResponse) {
    const journeys = apiResponse && apiResponse.journeys;
    if (!journeys || journeys.length === 0) return null;

    const journey = journeys[0];
    const departureDt = navitiaDatetimeToIso(journey.departure_date_time || "");
    const arrivalDt   = navitiaDatetimeToIso(journey.arrival_date_time || "");
    const durationMin = Math.round((journey.duration || 0) / 60);
    const nbTransfers = journey.nb_transfers || 0;
    const sections    = journey.sections || [];

    const fromNom = extractStationName((sections[0] || {}).from || {});
    const toNom   = extractStationName((sections[sections.length - 1] || {}).to || {});

    return {
      from_station_nom: fromNom,
      to_station_nom: toNom,
      departure_datetime: departureDt,
      arrival_datetime: arrivalDt,
      duration_minutes: durationMin,
      nb_transfers: nbTransfers,
      sections: parseJourneySections(journey),
    };
  }

  /**
   * Convert a duration in minutes to a human-readable French-style string.
   *
   * @param {number} totalMinutes - Non-negative integer number of minutes.
   * @returns {string} E.g. '1h 35min', '45min', '2h'.
   */
  function formatDurationMinutes(totalMinutes) {
    const hours   = Math.floor(totalMinutes / 60);
    const minutes = totalMinutes % 60;
    if (hours === 0) return `${minutes}min`;
    if (minutes === 0) return `${hours}h`;
    return `${hours}h ${minutes}min`;
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  window.InterJourney = {
    navitiaDatetimeToIso,
    extractStationName,
    parseJourneySections,
    parseBestJourney,
    formatDurationMinutes,
  };
})();
